packages/incremental/src/query.candy

```diff
@@ -0,0 +1,32 @@
+use ..context
+
+class Query<T> {
+  /// A query that can be run and produces a `T`.
+
+  let name: String
+  /// A unique query name.
+  /// 
+  /// The query name usually equals the name of the method or function that runs the query, possibly
+  /// with a package prefix. For example, you might have a `"math.computeSign"` query that computes
+  /// the sign of an `Int`.
+
+  let input: Equals & Hash
+  /// The input to the query.
+  ///
+  /// Multiple inputs can be represented as `Tuple`s or `List`s.
+
+  let runner: () => T
+  /// A runner that actually runs the query.
+  ///
+  /// The runner should be deterministic with regard to the inputs.
+
+  let shouldAlwaysRun: Bool
```

Comment:
Maybe we should rename this to a higher-level "isPure" or "isSideEffectFree". Also, this is currently ignored during execution.